### ISSUE TITLE
panic in istio-init with GKE+COS and interceptionMode: TPROXY

### ISSUE THREAD

**Issue by skolenkin** [istio/istio#22500]
GKE: v1.14.10-gke.24
 istio: 1.5.1 (Also I have tested on Istio 1.5.0 and I have got the same issue)
NGINX Deployment:
```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: nginx
spec:
  replicas: 1 # tells deployment to run 2 pods matching the template
  template: # create pods using pod definition in this template
    metadata:
      labels:
        app: nginx
        type: webserver
      annotations:
        sidecar.istio.io/interceptionMode: TPROXY
        sidecar.istio.io/proxyImage: docker.io/istio/proxyv2:1.5.0
    spec:
      containers:
      - name: nginx
        image: nginx:1.12.1
        ports:
        - containerPort: 80
          name: http-server

```

```
kubectl logs -f nginx-5964bfcd-9jxgd -c istio-init
Environment:
------------
ENVOY_PORT=
INBOUND_CAPTURE_PORT=
ISTIO_INBOUND_INTERCEPTION_MODE=
ISTIO_INBOUND_TPROXY_MARK=
ISTIO_INBOUND_TPROXY_ROUTE_TABLE=
ISTIO_INBOUND_PORTS=
ISTIO_LOCAL_EXCLUDE_PORTS=
ISTIO_SERVICE_CIDR=
ISTIO_SERVICE_EXCLUDE_CIDR=

Variables:
----------
PROXY_PORT=15001
PROXY_INBOUND_CAPTURE_PORT=15006
PROXY_UID=1337
PROXY_GID=1337
INBOUND_INTERCEPTION_MODE=TPROXY
INBOUND_TPROXY_MARK=1337
INBOUND_TPROXY_ROUTE_TABLE=133
INBOUND_PORTS_INCLUDE=80
INBOUND_PORTS_EXCLUDE=15090,15020
OUTBOUND_IP_RANGES_INCLUDE=*
OUTBOUND_IP_RANGES_EXCLUDE=
OUTBOUND_PORTS_EXCLUDE=
KUBEVIRT_INTERFACES=
ENABLE_INBOUND_IPV6=false

ip -f inet rule add fwmark 1337 lookup 133
ip -f inet route add local default dev lo table 133
RTNETLINK answers: File exists
ip route show table all
local default dev lo table 133 scope host
default via 10.24.0.1 dev eth0
10.24.0.0/24 via 10.24.0.1 dev eth0 src 10.24.0.12
10.24.0.1 dev eth0 scope link src 10.24.0.12
broadcast 10.24.0.0 dev eth0 table local proto kernel scope link src 10.24.0.12
local 10.24.0.12 dev eth0 table local proto kernel scope host src 10.24.0.12
broadcast 10.24.0.255 dev eth0 table local proto kernel scope link src 10.24.0.12
broadcast 127.0.0.0 dev lo table local proto kernel scope link src 127.0.0.1
local 127.0.0.0/8 dev lo table local proto kernel scope host src 127.0.0.1
local 127.0.0.1 dev lo table local proto kernel scope host src 127.0.0.1
broadcast 127.255.255.255 dev lo table local proto kernel scope link src 127.0.0.1
Writing following contents to rules file:  /tmp/iptables-rules-1585226543293516399.txt869048266
* mangle
-N ISTIO_DIVERT
-N ISTIO_TPROXY
-N ISTIO_INBOUND
-A ISTIO_DIVERT -j MARK --set-mark 1337
-A ISTIO_DIVERT -j ACCEPT
-A ISTIO_TPROXY ! -d 127.0.0.1/32 -p tcp -j TPROXY --tproxy-mark 1337/0xffffffff --on-port 15001
-A PREROUTING -p tcp -j ISTIO_INBOUND
-A ISTIO_INBOUND -p tcp --dport 80 -m socket -j ISTIO_DIVERT
-A ISTIO_INBOUND -p tcp --dport 80 -m socket -j ISTIO_DIVERT
-A ISTIO_INBOUND -p tcp --dport 80 -j ISTIO_TPROXY
COMMIT
* nat
-N ISTIO_REDIRECT
-N ISTIO_IN_REDIRECT
-N ISTIO_OUTPUT
-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-port 15001
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-port 15001
-A OUTPUT -p tcp -j ISTIO_OUTPUT
-A ISTIO_OUTPUT -o lo -s 127.0.0.6/32 -j RETURN
-A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -m owner --uid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -o lo -m owner ! --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -o lo ! -d 127.0.0.1/32 -m owner --gid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -o lo -m owner ! --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN
-A ISTIO_OUTPUT -j ISTIO_REDIRECT
COMMIT

iptables-restore --noflush /tmp/iptables-rules-1585226543293516399.txt869048266
iptables-restore: line 12 failed
iptables-save
# Generated by iptables-save v1.6.1 on Thu Mar 26 12:42:23 2020
*nat
:PREROUTING ACCEPT [43:2580]
:INPUT ACCEPT [43:2580]
:OUTPUT ACCEPT [0:0]
:POSTROUTING ACCEPT [0:0]
:ISTIO_IN_REDIRECT - [0:0]
:ISTIO_OUTPUT - [0:0]
:ISTIO_REDIRECT - [0:0]
-A OUTPUT -p tcp -j ISTIO_OUTPUT
-A ISTIO_IN_REDIRECT -p tcp -j REDIRECT --to-ports 15001
-A ISTIO_OUTPUT -s 127.0.0.6/32 -o lo -j RETURN
-A ISTIO_OUTPUT ! -d 127.0.0.1/32 -o lo -m owner --uid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -o lo -m owner ! --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -m owner --uid-owner 1337 -j RETURN
-A ISTIO_OUTPUT ! -d 127.0.0.1/32 -o lo -m owner --gid-owner 1337 -j ISTIO_IN_REDIRECT
-A ISTIO_OUTPUT -o lo -m owner ! --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -m owner --gid-owner 1337 -j RETURN
-A ISTIO_OUTPUT -d 127.0.0.1/32 -j RETURN
-A ISTIO_OUTPUT -j ISTIO_REDIRECT
-A ISTIO_REDIRECT -p tcp -j REDIRECT --to-ports 15001
COMMIT
# Completed on Thu Mar 26 12:42:23 2020
# Generated by iptables-save v1.6.1 on Thu Mar 26 12:42:23 2020
*mangle
:PREROUTING ACCEPT [48:2906]
:INPUT ACCEPT [45:2700]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [45:1800]
:POSTROUTING ACCEPT [45:1800]
COMMIT
# Completed on Thu Mar 26 12:42:23 2020
panic: exit status 1

goroutine 1 [running]:
istio.io/istio/tools/istio-iptables/pkg/dependencies.(*RealDependencies).RunOrFail(0xd819c0, 0x9739b8, 0x10, 0xc000098ca0, 0x2, 0x2)
	istio.io/istio@/tools/istio-iptables/pkg/dependencies/implementation.go:44 +0x96
istio.io/istio/tools/istio-iptables/pkg/cmd.(*IptablesConfigurator).executeIptablesRestoreCommand(0xc000101d30, 0x7ff4c8630d01, 0x0, 0x0)
	istio.io/istio@/tools/istio-iptables/pkg/cmd/run.go:474 +0x3aa
istio.io/istio/tools/istio-iptables/pkg/cmd.(*IptablesConfigurator).executeCommands(0xc000101d30)
	istio.io/istio@/tools/istio-iptables/pkg/cmd/run.go:481 +0x45
istio.io/istio/tools/istio-iptables/pkg/cmd.(*IptablesConfigurator).run(0xc000101d30)
	istio.io/istio@/tools/istio-iptables/pkg/cmd/run.go:428 +0x24e2
istio.io/istio/tools/istio-iptables/pkg/cmd.glob..func1(0xd5c740, 0xc0000e2700, 0x0, 0x10)
	istio.io/istio@/tools/istio-iptables/pkg/cmd/root.go:56 +0x14e
github.com/spf13/cobra.(*Command).execute(0xd5c740, 0xc0000aa010, 0x10, 0x11, 0xd5c740, 0xc0000aa010)
	github.com/spf13/cobra@v0.0.5/command.go:830 +0x2aa
github.com/spf13/cobra.(*Command).ExecuteC(0xd5c740, 0x40574f, 0xc00007e058, 0x0)
	github.com/spf13/cobra@v0.0.5/command.go:914 +0x2fb
github.com/spf13/cobra.(*Command).Execute(...)
	github.com/spf13/cobra@v0.0.5/command.go:864
istio.io/istio/tools/istio-iptables/pkg/cmd.Execute()
	istio.io/istio@/tools/istio-iptables/pkg/cmd/root.go:284 +0x2d
main.main()
	istio.io/istio@/tools/istio-iptables/main.go:22 +0x20
```

**Comment by howardjohn** [comment]
@abhide @rlenglet

**Comment by rlenglet** [comment]
Just to scope the problem better, from @jrajahalme's https://github.com/istio/istio/issues/19380#issuecomment-603527006 and this issue's description, `istio-init` fails to configure `mangle` table, which happens when all these conditions are met:

- GKE
- COS as the node system
- TPROXY interception mode

**Comment by rlenglet** [comment]
@jrajahalme investigated deeper and found that this is the rule that is causing the failure:
```
-A ISTIO_INBOUND -p tcp -m socket -j ISTIO_DIVERT
```

This is due to COS not shipping with the `xt_socket` kernel module.

@jrajahalme tried removing that rule, which eliminates the `istio-init` crash, but upstream connections fail at runtime.
We're trying to find an alternate implementation that won't require `iptables`'s `socket` module.

**Comment by jrajahalme** [comment]
@rlenglet Figured out what happens when `-m socket` rule is removed. Return packets from istio-pilot get dropped, so `connect()` calls to pilot xds fail. If I replace the `-m socket` rule with the following rule, connectivity to pilot succeeds:

```
[26:104118] -A ISTIO_INBOUND -p tcp -m tcp --sport 15011 -j RETURN
```
So it happens that the `TPROXY` rule is causing the return packets to be dropped IF the found established socket is not `transparent`:

```
        /* NOTE: assign_sock consumes our sk reference */
        if (sk && nf_tproxy_sk_is_transparent(sk)) {
                /* This should be in a separate target, but we don't do multiple
                   targets on the same rule yet */
                skb->mark = (skb->mark & ~mark_mask) ^ mark_value;
...
                nf_tproxy_assign_sock(skb, sk);
                return NF_ACCEPT;
        }
...
        return NF_DROP;
```

**Comment by jrajahalme** [comment]
Replacing the `-m socket` rule with a `-m conntrack` rule allows connections to pilot to succeed:
```
# Generated by iptables-save v1.6.1 on Thu Mar 26 19:13:51 2020
*mangle
:PREROUTING ACCEPT [370:41883]
:INPUT ACCEPT [370:41883]
:FORWARD ACCEPT [0:0]
:OUTPUT ACCEPT [356:39663]
:POSTROUTING ACCEPT [356:39663]
:ISTIO_DIVERT - [0:0]
:ISTIO_INBOUND - [0:0]
:ISTIO_TPROXY - [0:0]
[296466:33561855] -A PREROUTING -p tcp -j ISTIO_INBOUND
[0:0] -A ISTIO_DIVERT -j MARK --set-xmark 0x539/0xffffffff
[0:0] -A ISTIO_DIVERT -j ACCEPT
[0:0] -A ISTIO_INBOUND -p tcp -m tcp --dport 22 -j RETURN
[173111:13573977] -A ISTIO_INBOUND -p tcp -m tcp --dport 15020 -j RETURN
[150:25742] -A ISTIO_INBOUND -p tcp -m conntrack --ctstate RELATED,ESTABLISHED -j RETURN
[0:0] -A ISTIO_INBOUND -p tcp -j ISTIO_TPROXY
[23172:1390320] -A ISTIO_TPROXY ! -d 127.0.0.1/32 -p tcp -j TPROXY --on-port 15001 --on-ip 0.0.0.0 --tproxy-mark 0x539/0xffffffff
COMMIT
```
Not sure yet if everything works, but conceptually these should be equivalent when the destination IP address is a local address of the POD so no skb marking is needed for policy routing.

**Comment by jrajahalme** [comment]
In the above the `ISTIO_DIVERT` chain is no longer used, and the `TPROXY` rule could be placed directly in the ISTIO_INBOUND chain. I'll test a bit more and propose a PR.

**Comment by jrajahalme** [comment]
@rlenglet The above works for sidecars since incoming connections are for the POD IP, so policy routing is not needed. Do you know if TPROXY redirection is used for any proxy instances that are not sidecars?

**Comment by rlenglet** [comment]
Thanks a lot for the fix, @jrajahalme!!!

**Comment by rlenglet** [comment]
@jrajahalme's fix will go into the 1.5.2 release. Thanks @jrajahalme and @fpesce!

**Comment by longhoangwkm** [comment]
we got this problem on GKE now, can i know when this version has been released ?!

**Comment by rlenglet** [comment]
> we got this problem on GKE now, can i know when this version will be released ?!

@jrajahalme's fix will go into the 1.5.2 release.

**Comment by longhoangwkm** [comment]
Thanks, can't wait to see
